eth2/types/src/slot_epoch_root.rs

```
@@ -51,11 +53,53 @@ impl Slot {
 }
 
 impl Epoch {
+    pub fn new(epoch: u64) -> Self {
+        Self(epoch)
+    }
+}
+
+impl EEIndex {
+    pub fn new(ee_index: u64) -> Self {
+        Self(ee_index)
+    }
+}
+
+impl Shard {
+    pub fn new(shard: u64) -> Self {
+        Self(shard)
+    }
+}
+
+impl ShardSlot {
     pub fn new(slot: u64) -> Self {
         Self(slot)
     }
 }
 
+/// Convert from a hex string to a Root
+impl TryFrom<&str> for Root {
+    type Error = String;
+
+    fn try_from(s: &str) -> Result<Self, Self::Error> {
+        let vec = Vec::from_hex(s).map_err(|e| {
+            return format!("cannot convert string to hex: {:?}", e);
```

Converting an `Error` into a `String` loses a lot of contextual information. Generally it'd be better to either return the original error type, or convert it to this crate's error type if you don't want to leak dependencies.

---

Thanks for the callout!

This may be me being fairly new to Rust, but I'm not sure I totally agree here.  Certainly I agree that returning a string will make it so the underlying Error struct defined by the `hex` crate won't get passed along in the return value of `try_from`.  That said, I think all the relevant context will be there, since the debug print will just print out the underlying error from the `hex` crate and include it in the `try_from` error message.

I don't feel strongly here at all, but to me it doesn't seem compelling to have the extra logic to either 1) define an `Error` type for the `eth2/types` library or 2) just return the underlying error from `hex`.  Seems cleaner, simpler, and just as effective to have it as written.

Interested in reading your response on this :)